### PR TITLE
Remove polling from auth extensions

### LINKED ISSUE #108342: Multiple prompt to grant access to login keychain
- VSCode Version: 1.50-Insiders
- OS Version: macOS

@deepak1556 and me investigated but didn't find anything special from the logs `log show --start "2020-10-07 15:27:48" --end "2020-10-07 16:27:48" --predicate 'eventMessage CONTAINS[c] "Keychain"' --info --debug` and keychain info. One suspicion is maybe I'm using GitHub intensively, my setup:

* Settings Sync is using GitHub account
* GitHub PRI is using GitHub account
* GitHub Issue Notebook is also using GitHub account

On Oct 6th, I ran into SSO bad credential issue with GitHub Issue Notebook so I signed out of my GitHub account. Then I login again and grant access to all three extensions/services. I got a prompt for login keychain access once, I clicked "always allow"

On Oct 7th afternoon, I saw the prompt again.

Does this issue occur when all extensions are disabled?: Yes/No

### PR DESCRIPTION
**Pull request by RMacfarlane**
This PR fixes #108342

### CODE REVIEW

File: extensions/github-authentication/src/common/keychain.ts [hunk 1]
```diff
@@ -76,6 +76,11 @@ export class Keychain {
 			return Promise.resolve(undefined);
 		}
 	}
+	onDidChangePassword(listener: () => void) {
```
Comment (bpasero) [inline extensions/github-authentication/src/common/keychain.ts:79]: Not clear to me why this indirection is needed and `vscode.authentication.onDidChangePassword` cannot be used directly.

File: extensions/github-authentication/src/github.ts [hunk 2]
```diff
@@ -33,52 +33,48 @@ export class GitHubAuthenticationProvider {
 			// Ignore, network request failed
 		}
 
-		this.pollForChange();
+		keychain.onDidChangePassword(() => this.checkForUpdates());
```
Comment (bpasero) [inline extensions/github-authentication/src/github.ts:36]: Does this listener need to be disposed with any lifecycle?

File: extensions/microsoft-authentication/src/AADHelper.ts [hunk 3]
```diff
@@ -140,7 +140,7 @@ export class AzureActiveDirectoryService {
 			}
 		}
 
-		this.pollForChange();
+		keychain.onDidChangePassword(() => this.checkForUpdates);
```
Comment (bpasero) [inline extensions/microsoft-authentication/src/AADHelper.ts:143]: Does this listener need to be disposed with any lifecycle?

File: extensions/microsoft-authentication/src/keychain.ts [hunk 4]
```diff
@@ -85,6 +85,11 @@ export class Keychain {
 			return Promise.resolve(undefined);
 		}
 	}
+	onDidChangePassword(listener: () => void) {
```
Comment (bpasero) [inline extensions/microsoft-authentication/src/keychain.ts:88]: Not clear to me why this indirection is needed and `vscode.authentication.onDidChangePassword` cannot be used directly.

File: src/vs/workbench/api/common/extHostAuthentication.ts [hunk 5]
```diff
@@ -24,6 +24,10 @@ export class ExtHostAuthentication implements ExtHostAuthenticationShape {
 	private _onDidChangeSessions = new Emitter<vscode.AuthenticationSessionsChangeEvent>();
 	readonly onDidChangeSessions: Event<vscode.AuthenticationSessionsChangeEvent> = this._onDidChangeSessions.event;
 
+
```
Comment (bpasero) [inline src/vs/workbench/api/common/extHostAuthentication.ts:27]: Extra newline

File: src/vs/workbench/api/common/extHostAuthentication.ts [hunk 6]
```diff
@@ -203,4 +207,8 @@ export class ExtHostAuthentication implements ExtHostAuthenticationShape {
 		this._onDidChangeAuthenticationProviders.fire({ added, removed });
 		return Promise.resolve();
 	}
+	$onDidChangePassword(): Promise<void> {
```
Comment (bpasero) [inline src/vs/workbench/api/common/extHostAuthentication.ts:210]: Missing newline :)

File: src/vs/workbench/api/common/extHostAuthentication.ts [hunk 7]
```diff
@@ -203,4 +207,8 @@ export class ExtHostAuthentication implements ExtHostAuthenticationShape {
...
+		this._onDidChangePassword.fire();
+		return Promise.resolve();
```
Comment (bpasero) [inline src/vs/workbench/api/common/extHostAuthentication.ts:212]: If you use `async` on the method, you don't have to return a promise like this.

File: src/vs/workbench/services/credentials/browser/credentialsService.ts [hunk 8]
```diff
@@ -6,11 +6,15 @@
...
 import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
 import { IWorkbenchEnvironmentService } from 'vs/workbench/services/environment/common/environmentService';
+import { Emitter } from 'vs/base/common/event';
 
 export class BrowserCredentialsService implements ICredentialsService {
 
 	declare readonly _serviceBrand: undefined;
 
+	private _onDidChangePassword: Emitter<void> = new Emitter();
+	onDidChangePassword = this._onDidChangePassword.event;
```
Comment (bpasero) [inline src/vs/workbench/services/credentials/browser/credentialsService.ts:16]:
```suggestion
	readonly onDidChangePassword = this._onDidChangePassword.event;
```

File: src/vs/workbench/services/credentials/common/credentials.ts [hunk 9]
```diff
@@ -17,4 +18,5 @@ export interface ICredentialsProvider {
 
 export interface ICredentialsService extends ICredentialsProvider {
 	readonly _serviceBrand: undefined;
+	onDidChangePassword: Event<void>;
```
Comment (bpasero) [inline src/vs/workbench/services/credentials/common/credentials.ts:21]:
```suggestion
	readonly onDidChangePassword: Event<void>;
```

**Review by bpasero — APPROVED** [review]
@RMacfarlane looks really good overall. I left some :lipstick: and lifecycle related review comments that we can do on `master`. I typically tend to let a thing `extends Disposable` whenever I deal with events or emitters to register them properly to get disposed. Even if the outer thing is never getting disposed, at least inside disposables are tracked properly. I can do this change on `master` as well next week.

**Comment by bpasero** [comment]
@RMacfarlane after a bit more thinking, should the event possibly be more clever around this:
* only fire if the password actually changed
* carry the password around as part of the event

The reason I ask is that very likely a client of this event will ask for the new password, and that would trigger another roundtrip through all the processes until resolved. If in `setPassword` we would call `getPassword` to check if the password actually changed and add it to the event, wouldn't that make things easier?

The only downside I can think of is that `getPassword` would trigger potentially a dialog on macOS unless the user has granted this access already permanently.

**Comment by RMacfarlane** [comment]
Thanks for reviewing! I've pushed changes addressing your feedback to master and also here.

I think we should avoid directly passing the password as part of the change event, but I agree we could do more to avoid unnecessary round trips. My concern with adding the password in the change event is that this increases its exposure - we would need to also check that the listening extension has been permitted to access this. Generally though, this event should be fired infrequently, only when a user is actually signing in or out.

**Pull request merged**